## Нужно скачать обязательно

In [45]:
import sys, subprocess
def ensure(pkgs): subprocess.check_call([sys.executable, "-m", "pip", "install", "--quiet", *pkgs])
ensure(["ultralytics==8.3.129", "roboflow", "albumentations"])

## Библиотечки

In [2]:
import os
import random
import time
import ultralytics
from pathlib import Path

import numpy as np
import torch
from torch import nn
import yaml
from ultralytics import YOLO
from ultralytics.utils.metrics import box_iou
from ultralytics.utils.torch_utils import de_parallel
from roboflow import Roboflow

SEED = 42
torch.manual_seed(SEED)
np.random.seed(SEED)
random.seed(SEED)

DEVICE = 0 if torch.cuda.is_available() else 'cpu'
print('DEVICE:', torch.cuda.get_device_name(0) if DEVICE==0 else 'CPU')

PROJECT = 'ABOBUS'
IMG_SIZE = 512
EPOCHS_BS = 20
EPOCHS_IMPROVED = 24

WARNING ⚠️ user config directory '/root/.config/Ultralytics' is not writeable, defaulting to '/tmp' or CWD.Alternatively you can define a YOLO_CONFIG_DIR environment variable for this path.
Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/tmp/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
DEVICE: NVIDIA A100-PCIE-40GB


## Датасетик

### Выбор датасета

- **Практическая задача безопасности**: детекция защитных касок на рабочих площадках помогает автоматически контролировать соблюдение техники безопасности.  
- **Готовый и разнообразный**: Roboflow-проект содержит кадры из разных ракурсов, условий освещения и фоновых сцен — это улучшает обобщающую способность модели.  
- **YOLOv8-совместимый формат**: скачиваем архив уже с `data.yaml` и аннотациями в нужном формате, что упрощает интеграцию с Ultralytics YOLO-пайплайном.  
- **Лёгкая загрузка и версияция**: Roboflow API обеспечивает версионирование данных (здесь version 3), что гарантирует воспроизводимость эксперимента.  


In [3]:
from pathlib import Path
from zipfile import ZipFile
from roboflow import Roboflow

rf       = Roboflow(api_key="57PY7WAqNZUOUmf68ywL")
project  = rf.workspace("learning-evidence").project("helmet-detection_yolov8")
version  = project.version(3)
dataset  = version.download("yolov8")

zip_path   = Path(dataset.location) / f"{project.name}-{version.version}.zip"
extract_to = Path(dataset.location)

if zip_path.exists() and not (extract_to / "data.yaml").exists():
    print("Unzipping dataset...")
    with ZipFile(zip_path, 'r') as zf:
        zf.extractall(extract_to)
    print("Done")

DATA_YAML = extract_to / "data.yaml"
print("DATA_YAML →", DATA_YAML)

loading Roboflow workspace...
loading Roboflow project...



Extracting Dataset Version Zip to Helmet-Detection_YOLOv8-3 in yolov8:: 100%|██████████| 8350/8350 [00:01<00:00, 5437.87it/s]


DATA_YAML → /home/Helmet-Detection_YOLOv8-3/data.yaml


## Очень важная функция

### Почему выбираем mAP@50 и mAP@50–95

- **mAP@50 (PASCAL VOC-style)**  
  - Использует порог перекрытия IoU = 0.5, то есть относительно «мягкий» критерий: объект считается найденным, если предсказанный бокс перекрывает граничную правду наполовину.  
  - Хорошо показывает базовые способности модели к детекции и классификации без излишней строгости по точности локализации.

- **mAP@50–95 (COCO-style)**  
  - Усредняет mAP по IoU-порогам от 0.5 до 0.95 с шагом


In [4]:
def train_and_eval(weights, name, epochs, imgsz=IMG_SIZE, lr=0.01,
                   batch=16, aug=None):
    model = YOLO(weights)
    results = model.train(data=str(DATA_YAML),
                          epochs=epochs, imgsz=imgsz,
                          batch=batch, lr0=lr,
                          device=DEVICE,
                          project=PROJECT, name=name,
                          amp=True, seed=SEED,
                          **(aug or {}))
    val_res = model.val(data=str(DATA_YAML), split='test', device=DEVICE)
    m50, m5095 = val_res.box.map50, val_res.box.map
    print(f"{name:14s} ➜ mAP50 {m50:.3f} | mAP50-95 {m5095:.3f}")
    return model, (m50, m5095)

## Baseline

## Обычный

- Модель: YOLOv8n (nano), компактная CNN-версия детектора.  
- Датасет: Roboflow “helmet-detection” (2 класса: “With Helmet” / “Without Helmet”).  
- Тренинг: 20 эпох, imgsz 512, batch 32, lr₀ 0.01 (auto→0.00167), AdamW, AMP, patience 100.  

In [5]:
y8n, metrics_y8n = train_and_eval('yolov8n.pt', 'y8n_baseline',
                                  epochs=EPOCHS_BS, batch=32, lr=0.01)

100%|██████████| 6.25M/6.25M [00:00<00:00, 20.5MB/s]


New https://pypi.org/project/ultralytics/8.3.130 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=32, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/home/Helmet-Detection_YOLOv8-3/data.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=20, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=512, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=y8n_baseline, nbs=64, nms=Fal

100%|██████████| 755k/755k [00:00<00:00, 5.25MB/s]

Overriding model.yaml nc=80 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics

 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]  

[W NNPACK.cpp:61] Could not initialize NNPACK! Reason: Unsupported hardware.


Model summary: 129 layers, 3,011,238 parameters, 3,011,222 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks...


100%|██████████| 5.35M/5.35M [00:00<00:00, 22.4MB/s]


AMP: checks passed ✅
train: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1353.7±390.3 MB/s, size: 63.9 KB)


train: Scanning /home/Helmet-Detection_YOLOv8-3/train/labels... 3648 images, 13 backgrounds, 0 corrupt: 100%|██████████| 3648/3648 [00:03<00:00, 1073.69it/s]

train: New cache created: /home/Helmet-Detection_YOLOv8-3/train/labels.cache


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 707.5±496.3 MB/s, size: 77.5 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/valid/labels... 372 images, 1 backgrounds, 0 corrupt: 100%|██████████| 372/372 [00:00<00:00, 1207.54it/s]

val: New cache created: /home/Helmet-Detection_YOLOv8-3/valid/labels.cache


Plotting labels to ABOBUS/y8n_baseline/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 512 train, 512 val
Using 6 dataloader workers
Logging results to ABOBUS/y8n_baseline
Starting training for 20 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/20      2.74G      1.579       2.16      1.244        134        512: 100%|██████████| 114/114 [00:42<00:00,  2.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:11<00:00,  1.87s/it]

                   all        372        660      0.708      0.523      0.602      0.362



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/20      3.11G      1.516       1.38      1.215         85        512: 100%|██████████| 114/114 [00:37<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.14it/s]

                   all        372        660      0.746      0.731      0.796      0.465



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/20      3.12G      1.491        1.2      1.195         53        512: 100%|██████████| 114/114 [00:26<00:00,  4.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.58s/it]

                   all        372        660      0.734      0.744      0.772      0.444



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/20      3.12G      1.464      1.118      1.198         98        512: 100%|██████████| 114/114 [00:38<00:00,  2.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.53s/it]

                   all        372        660      0.792      0.682      0.777      0.447



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/20      3.12G      1.447      1.043      1.194        101        512: 100%|██████████| 114/114 [00:39<00:00,  2.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.58s/it]

                   all        372        660      0.809      0.802      0.849      0.491



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/20      3.12G      1.405     0.9626      1.167         92        512: 100%|██████████| 114/114 [00:28<00:00,  4.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:02<00:00,  2.84it/s]

                   all        372        660      0.834       0.79      0.862      0.525



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/20      3.12G      1.391     0.9369      1.166        103        512: 100%|██████████| 114/114 [00:36<00:00,  3.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.55s/it]

                   all        372        660      0.837      0.812      0.874      0.539



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/20      3.12G      1.365      0.895      1.137        116        512: 100%|██████████| 114/114 [00:37<00:00,  3.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.54s/it]

                   all        372        660      0.873      0.816        0.9      0.565



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/20      3.12G      1.343     0.8477      1.136         93        512: 100%|██████████| 114/114 [00:37<00:00,  3.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.55s/it]

                   all        372        660      0.868      0.843      0.913      0.559



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/20      3.12G       1.34     0.8373      1.128         88        512: 100%|██████████| 114/114 [00:19<00:00,  5.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:02<00:00,  2.91it/s]

                   all        372        660       0.86      0.847      0.914      0.572


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/20      3.12G      1.286     0.7578      1.128         65        512: 100%|██████████| 114/114 [00:20<00:00,  5.61it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.23it/s]

                   all        372        660      0.805      0.833      0.865      0.529



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/20      3.12G      1.266      0.722      1.123         71        512: 100%|██████████| 114/114 [00:19<00:00,  5.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.35it/s]

                   all        372        660      0.875      0.838      0.912      0.564



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/20      3.12G      1.227     0.7001      1.111         65        512: 100%|██████████| 114/114 [00:19<00:00,  5.82it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.21it/s]

                   all        372        660      0.863      0.849      0.919      0.584



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/20      3.12G      1.207     0.6681      1.093         64        512: 100%|██████████| 114/114 [00:19<00:00,  5.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.55it/s]

                   all        372        660      0.875      0.813      0.914      0.583



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/20      3.12G      1.184     0.6477      1.081         59        512: 100%|██████████| 114/114 [00:20<00:00,  5.64it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.40it/s]

                   all        372        660      0.847      0.882      0.914      0.571



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/20      3.12G      1.146     0.6124      1.065         65        512: 100%|██████████| 114/114 [00:20<00:00,  5.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.32it/s]

                   all        372        660      0.842      0.878      0.914      0.574



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/20      3.12G      1.125     0.5866      1.055         55        512: 100%|██████████| 114/114 [00:19<00:00,  5.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.25it/s]

                   all        372        660      0.878      0.894      0.934      0.593



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/20      3.12G      1.096     0.5641       1.04         56        512: 100%|██████████| 114/114 [00:20<00:00,  5.55it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:02<00:00,  2.86it/s]

                   all        372        660      0.879      0.882      0.932      0.589



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/20      3.12G       1.06     0.5368      1.028         65        512: 100%|██████████| 114/114 [00:18<00:00,  6.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.70it/s]

                   all        372        660      0.868       0.89      0.929      0.595



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/20      3.12G      1.035     0.5191      1.018         57        512: 100%|██████████| 114/114 [00:17<00:00,  6.58it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  3.30it/s]

                   all        372        660      0.879      0.876      0.931      0.596



20 epochs completed in 0.178 hours.
Optimizer stripped from ABOBUS/y8n_baseline/weights/last.pt, 6.2MB
Optimizer stripped from ABOBUS/y8n_baseline/weights/best.pt, 6.2MB

Validating ABOBUS/y8n_baseline/weights/best.pt...
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
Model summary (fused): 72 layers, 3,006,038 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:03<00:00,  1.72it/s]


                   all        372        660      0.879      0.876      0.931      0.596
           With Helmet        215        338       0.89      0.917      0.957      0.662
        Without Helmet        208        322      0.868      0.835      0.905       0.53
Speed: 0.2ms preprocess, 0.4ms inference, 0.0ms loss, 3.2ms postprocess per image
Results saved to ABOBUS/y8n_baseline
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
Model summary (fused): 72 layers, 3,006,038 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 816.6±321.0 MB/s, size: 69.7 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/test/labels... 149 images, 0 backgrounds, 0 corrupt: 100%|██████████| 149/149 [00:00<00:00, 1658.26it/s]

val: /home/Helmet-Detection_YOLOv8-3/test/images/BikesHelmets446_png.rf.1dc30fe7d9690524fe33ffc1ee665fac.jpg: 1 duplicate labels removed
val: New cache created: /home/Helmet-Detection_YOLOv8-3/test/labels.cache



                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 5/5 [00:04<00:00,  1.11it/s]


                   all        149        273      0.897      0.859      0.913      0.582
           With Helmet        110        170      0.869      0.888      0.921      0.628
        Without Helmet         61        103      0.924       0.83      0.904      0.536
Speed: 2.7ms preprocess, 7.0ms inference, 0.0ms loss, 4.6ms postprocess per image
Results saved to ABOBUS/y8n_baseline2
y8n_baseline   ➜ mAP50 0.913 | mAP50-95 0.582


- **Высокая точность (0.897)**: мало ложных срабатываний.  
- **Немного ниже полнота (0.859)**: часть шлемов пропускается.  
- **Разрыв mAP50→mAP50-95** говорит о том, что локализация ещё не идеальна на строгих IoU.  

### Трансформер

- Модель: YOLOv8x, крупная CNN-архитектура (~68.2 M параметров, 257.4 GFLOPs).  
- Данные: тот же Roboflow “helmet-detection” (2 класса: “With Helmet” / “Without Helmet”).  
- Тренинг: 20 эпох, imgsz 512, batch 32, lr₀ 0.005→auto≈0.00167, AdamW, AMP, patience 100. 

In [6]:
r18, metrics_r18 = train_and_eval('yolov8x.pt', 'y8x_attn_base',
                                  epochs=EPOCHS_BS, batch=32, lr=0.005)

100%|██████████| 131M/131M [00:01<00:00, 78.2MB/s] 


New https://pypi.org/project/ultralytics/8.3.130 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/home/Helmet-Detection_YOLOv8-3/data.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=20, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=512, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8x.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=y8x_attn_base, nbs=64, nms=Fa

train: Scanning /home/Helmet-Detection_YOLOv8-3/train/labels.cache... 3648 images, 13 backgrounds, 0 corrupt: 100%|██████████| 3648/3648 [00:00<?, ?it/s]


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 742.5±143.9 MB/s, size: 77.5 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/valid/labels.cache... 372 images, 1 backgrounds, 0 corrupt: 100%|██████████| 372/372 [00:00<?, ?it/s]


Plotting labels to ABOBUS/y8x_attn_base/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.005' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 512 train, 512 val
Using 6 dataloader workers
Logging results to ABOBUS/y8x_attn_base
Starting training for 20 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/20      5.05G      1.587      1.426      1.491         25        512: 100%|██████████| 456/456 [01:33<00:00,  4.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:04<00:00,  4.87it/s]

                   all        372        660      0.639      0.672      0.742      0.409



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/20      5.68G      1.605      1.294      1.526         23        512: 100%|██████████| 456/456 [01:19<00:00,  5.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:04<00:00,  5.12it/s]

                   all        372        660      0.743      0.732      0.785      0.432



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/20      5.71G      1.561      1.181      1.476         26        512: 100%|██████████| 456/456 [01:22<00:00,  5.50it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:04<00:00,  5.96it/s]

                   all        372        660       0.79      0.768      0.821      0.473



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/20      5.71G      1.511      1.078      1.447         19        512: 100%|██████████| 456/456 [01:21<00:00,  5.62it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:04<00:00,  5.45it/s]

                   all        372        660      0.782      0.792      0.838        0.5



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/20      5.71G       1.48      1.049      1.424         21        512: 100%|██████████| 456/456 [01:20<00:00,  5.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:02<00:00,  8.05it/s]


                   all        372        660      0.802      0.793      0.855      0.507

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/20      5.71G      1.448     0.9879      1.407         29        512: 100%|██████████| 456/456 [01:24<00:00,  5.38it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:04<00:00,  5.00it/s]

                   all        372        660        0.8      0.829       0.88      0.539



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/20      5.71G      1.407     0.9108      1.377         30        512: 100%|██████████| 456/456 [01:14<00:00,  6.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:02<00:00,  8.19it/s]

                   all        372        660      0.827       0.84      0.886       0.54



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/20      5.71G       1.37     0.8739      1.342         17        512: 100%|██████████| 456/456 [02:07<00:00,  3.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:07<00:00,  3.03it/s]

                   all        372        660      0.827       0.84      0.889       0.55



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/20      5.71G      1.332     0.8255      1.322         28        512: 100%|██████████| 456/456 [01:53<00:00,  4.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:08<00:00,  2.98it/s]

                   all        372        660      0.819       0.86      0.904      0.556



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/20      5.71G      1.313     0.7963      1.312         11        512: 100%|██████████| 456/456 [01:53<00:00,  4.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:09<00:00,  2.60it/s]

                   all        372        660      0.851      0.847      0.902      0.565


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/20      5.71G      1.283     0.7274       1.37         12        512: 100%|██████████| 456/456 [02:05<00:00,  3.64it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:03<00:00,  7.32it/s]

                   all        372        660      0.842      0.848      0.905      0.573



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/20      5.71G      1.241     0.6836      1.335         19        512: 100%|██████████| 456/456 [02:14<00:00,  3.39it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:09<00:00,  2.46it/s]

                   all        372        660      0.844      0.873      0.908      0.578



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/20      5.71G       1.21     0.6473      1.325         19        512: 100%|██████████| 456/456 [01:29<00:00,  5.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:06<00:00,  3.60it/s]

                   all        372        660      0.871      0.844      0.912       0.58



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/20      5.71G      1.167     0.6129      1.287         13        512: 100%|██████████| 456/456 [01:53<00:00,  4.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:02<00:00,  8.08it/s]

                   all        372        660      0.823      0.858      0.906      0.579



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/20      5.71G      1.124     0.5834      1.264         16        512: 100%|██████████| 456/456 [01:53<00:00,  4.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:06<00:00,  3.52it/s]

                   all        372        660      0.853      0.862      0.912      0.582



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/20      5.71G      1.094     0.5405      1.248         12        512: 100%|██████████| 456/456 [01:54<00:00,  4.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:07<00:00,  3.19it/s]

                   all        372        660      0.858      0.877      0.924      0.591



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/20      5.71G      1.056     0.5227      1.215         10        512: 100%|██████████| 456/456 [01:42<00:00,  4.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:07<00:00,  3.28it/s]

                   all        372        660      0.895      0.828      0.924      0.589



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/20      5.71G      1.005      0.497      1.187         13        512: 100%|██████████| 456/456 [01:53<00:00,  4.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:07<00:00,  3.21it/s]

                   all        372        660      0.896       0.84      0.923      0.586



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/20      5.71G     0.9544     0.4663      1.158         14        512: 100%|██████████| 456/456 [01:53<00:00,  4.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:04<00:00,  5.01it/s]


                   all        372        660      0.874       0.88      0.933      0.591

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/20      5.71G     0.9178     0.4435      1.136         13        512: 100%|██████████| 456/456 [01:49<00:00,  4.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:07<00:00,  3.37it/s]

                   all        372        660      0.877       0.88      0.933      0.595



20 epochs completed in 0.633 hours.
Optimizer stripped from ABOBUS/y8x_attn_base/weights/last.pt, 136.7MB
Optimizer stripped from ABOBUS/y8x_attn_base/weights/best.pt, 136.7MB

Validating ABOBUS/y8x_attn_base/weights/best.pt...
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
Model summary (fused): 112 layers, 68,125,494 parameters, 0 gradients, 257.4 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 24/24 [00:08<00:00,  2.94it/s]


                   all        372        660      0.877       0.88      0.933      0.596
           With Helmet        215        338      0.912      0.917      0.962      0.669
        Without Helmet        208        322      0.842      0.843      0.904      0.522
Speed: 0.3ms preprocess, 4.8ms inference, 0.0ms loss, 5.1ms postprocess per image
Results saved to ABOBUS/y8x_attn_base
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
Model summary (fused): 112 layers, 68,125,494 parameters, 0 gradients, 257.4 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1990.1±374.1 MB/s, size: 69.7 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/test/labels.cache... 149 images, 0 backgrounds, 0 corrupt: 100%|██████████| 149/149 [00:00<?, ?it/s]

val: /home/Helmet-Detection_YOLOv8-3/test/images/BikesHelmets446_png.rf.1dc30fe7d9690524fe33ffc1ee665fac.jpg: 1 duplicate labels removed



                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 19/19 [00:04<00:00,  4.23it/s]


                   all        149        273      0.866      0.819      0.888      0.577
           With Helmet        110        170      0.881      0.859      0.916      0.638
        Without Helmet         61        103      0.852       0.78      0.861      0.516
Speed: 1.0ms preprocess, 9.5ms inference, 0.0ms loss, 6.1ms postprocess per image
Results saved to ABOBUS/y8x_attn_base2
y8x_attn_base  ➜ mAP50 0.888 | mAP50-95 0.577


- **Большая модель не улучшила mAP50** (0.888 vs 0.913 у YOLOv8n) — вероятно, переобучение или слишком высокий lr.  
- Незначительный рост mAP50-95 при снижении Precision/Recall → требуется доработка аугментаций и регуляризации.  

## Улучшенный baseline

In [7]:
strong_aug = dict(
    hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=15,
    translate=0.1, scale=0.7, shear=0.0, perspective=0.0005,
    mosaic=1.0, mixup=0.2, flipud=0.0, fliplr=0.5,
)

### Обычный++

**Архитектура:**  
Компактная CNN-средняя версия YOLOv8 (m) с агрессивными аугментациями (mosaic, mixup, цветовые преобразования и геометрия).

**Данные:**  
Тот же Roboflow “helmet-detection” (2 класса: “With Helmet” / “Without Helmet”).

**Тренинг:**  
24 эпохи, imgsz 512, batch 32, lr₀ 0.01→auto≈0.00167, AdamW, AMP, patience 100, strong_aug.

In [8]:
y8m, metrics_y8m = train_and_eval('yolov8m.pt', 'y8m_strong',
                                  epochs=EPOCHS_IMPROVED, batch=32,
                                  lr=0.01, aug=strong_aug)

100%|██████████| 49.7M/49.7M [00:00<00:00, 67.2MB/s]


New https://pypi.org/project/ultralytics/8.3.130 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/home/Helmet-Detection_YOLOv8-3/data.yaml, degrees=15, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=24, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=512, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.2, mode=train, model=yolov8m.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=y8m_strong, nbs=64, nms=False,

train: Scanning /home/Helmet-Detection_YOLOv8-3/train/labels.cache... 3648 images, 13 backgrounds, 0 corrupt: 100%|██████████| 3648/3648 [00:00<?, ?it/s]


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1390.8±811.7 MB/s, size: 77.5 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/valid/labels.cache... 372 images, 1 backgrounds, 0 corrupt: 100%|██████████| 372/372 [00:00<?, ?it/s]


Plotting labels to ABOBUS/y8m_strong/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 77 weight(decay=0.0), 84 weight(decay=0.0005), 83 bias(decay=0.0)
Image sizes 512 train, 512 val
Using 6 dataloader workers
Logging results to ABOBUS/y8m_strong
Starting training for 24 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/24      5.87G      1.696      1.669      1.449         64        512: 100%|██████████| 228/228 [00:48<00:00,  4.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:04<00:00,  2.68it/s]

                   all        372        660      0.666      0.534      0.592      0.314



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/24      6.42G       1.69      1.441       1.47         57        512: 100%|██████████| 228/228 [00:52<00:00,  4.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.47it/s]

                   all        372        660      0.766        0.7      0.758       0.34



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/24      6.42G      1.671      1.424      1.477         69        512: 100%|██████████| 228/228 [00:32<00:00,  6.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.88it/s]

                   all        372        660      0.785      0.717      0.798        0.4



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/24      6.44G      1.647      1.351       1.46         68        512: 100%|██████████| 228/228 [00:33<00:00,  6.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.91it/s]

                   all        372        660      0.799      0.769      0.836      0.436



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/24      6.44G      1.604      1.285      1.429         83        512: 100%|██████████| 228/228 [00:31<00:00,  7.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.48it/s]

                   all        372        660      0.802      0.798      0.858      0.425



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/24      6.44G      1.561      1.194      1.403         39        512: 100%|██████████| 228/228 [00:32<00:00,  7.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.02it/s]

                   all        372        660      0.783      0.825      0.875      0.489



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/24      6.44G      1.529      1.158      1.404         67        512: 100%|██████████| 228/228 [00:31<00:00,  7.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.45it/s]

                   all        372        660      0.842      0.793      0.875      0.501



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/24      6.44G       1.52      1.112      1.376         39        512: 100%|██████████| 228/228 [00:31<00:00,  7.27it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.04it/s]

                   all        372        660        0.8       0.83      0.877      0.482



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/24      6.44G      1.493      1.089      1.376         85        512: 100%|██████████| 228/228 [00:32<00:00,  7.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.85it/s]

                   all        372        660      0.836      0.824      0.902       0.53



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/24      6.44G      1.469       1.04      1.347         43        512: 100%|██████████| 228/228 [00:31<00:00,  7.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.15it/s]

                   all        372        660      0.842      0.783      0.873      0.525



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/24      6.44G      1.452      1.017      1.343         48        512: 100%|██████████| 228/228 [00:31<00:00,  7.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.92it/s]

                   all        372        660      0.895      0.804      0.917       0.52



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/24      6.44G      1.448     0.9867      1.342         54        512: 100%|██████████| 228/228 [00:32<00:00,  6.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.85it/s]

                   all        372        660      0.795      0.844      0.897      0.553



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/24      6.44G      1.419     0.9628      1.327         53        512: 100%|██████████| 228/228 [00:32<00:00,  6.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.05it/s]

                   all        372        660      0.827      0.842      0.915      0.536



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/24      6.44G      1.397     0.9366      1.301         61        512: 100%|██████████| 228/228 [00:31<00:00,  7.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.93it/s]

                   all        372        660       0.87       0.82      0.918      0.555


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/24      6.44G      1.295     0.7784      1.366         33        512: 100%|██████████| 228/228 [00:32<00:00,  7.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.81it/s]

                   all        372        660      0.861      0.828      0.913      0.551



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/24      6.44G      1.287     0.7632       1.37         27        512: 100%|██████████| 228/228 [00:31<00:00,  7.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.54it/s]

                   all        372        660      0.843      0.865      0.919      0.562



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/24      6.44G      1.262     0.7198      1.346         26        512: 100%|██████████| 228/228 [00:30<00:00,  7.37it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.75it/s]

                   all        372        660      0.843      0.875      0.932      0.564



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/24      6.44G      1.252     0.6998      1.343         26        512: 100%|██████████| 228/228 [00:30<00:00,  7.41it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.17it/s]

                   all        372        660      0.865       0.85      0.912      0.553



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/24      6.44G      1.217     0.6742      1.308         30        512: 100%|██████████| 228/228 [00:30<00:00,  7.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.81it/s]

                   all        372        660       0.86      0.876      0.928      0.586



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/24      6.44G      1.195     0.6467      1.299         29        512: 100%|██████████| 228/228 [00:31<00:00,  7.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.95it/s]

                   all        372        660      0.863      0.865      0.927      0.553



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/24      6.44G      1.181     0.6338      1.299         25        512: 100%|██████████| 228/228 [00:30<00:00,  7.39it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:01<00:00,  6.24it/s]

                   all        372        660       0.89      0.873      0.936      0.591



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/24      6.44G      1.141     0.6002      1.277         37        512: 100%|██████████| 228/228 [00:29<00:00,  7.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.72it/s]

                   all        372        660      0.884      0.878      0.939      0.589



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/24      6.44G      1.121     0.5826      1.259         26        512: 100%|██████████| 228/228 [00:32<00:00,  7.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.74it/s]

                   all        372        660      0.855      0.912      0.937      0.601



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/24      6.44G      1.092     0.5555      1.244         26        512: 100%|██████████| 228/228 [00:31<00:00,  7.27it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.91it/s]

                   all        372        660      0.879      0.893      0.939      0.593



24 epochs completed in 0.248 hours.
Optimizer stripped from ABOBUS/y8m_strong/weights/last.pt, 52.0MB
Optimizer stripped from ABOBUS/y8m_strong/weights/best.pt, 52.0MB

Validating ABOBUS/y8m_strong/weights/best.pt...
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
Model summary (fused): 92 layers, 25,840,918 parameters, 0 gradients, 78.7 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:03<00:00,  3.27it/s]


                   all        372        660      0.855      0.912      0.937      0.602
           With Helmet        215        338      0.837       0.95      0.964      0.668
        Without Helmet        208        322      0.873      0.875      0.911      0.536
Speed: 0.3ms preprocess, 1.8ms inference, 0.0ms loss, 1.7ms postprocess per image
Results saved to ABOBUS/y8m_strong
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
Model summary (fused): 92 layers, 25,840,918 parameters, 0 gradients, 78.7 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1401.4±717.7 MB/s, size: 69.7 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/test/labels.cache... 149 images, 0 backgrounds, 0 corrupt: 100%|██████████| 149/149 [00:00<?, ?it/s]

val: /home/Helmet-Detection_YOLOv8-3/test/images/BikesHelmets446_png.rf.1dc30fe7d9690524fe33ffc1ee665fac.jpg: 1 duplicate labels removed



                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:03<00:00,  3.25it/s]


                   all        149        273      0.891      0.829      0.906      0.596
           With Helmet        110        170      0.892      0.871      0.921      0.638
        Without Helmet         61        103      0.891      0.786      0.892      0.554
Speed: 1.2ms preprocess, 5.1ms inference, 0.0ms loss, 2.1ms postprocess per image
Results saved to ABOBUS/y8m_strong2
y8m_strong     ➜ mAP50 0.906 | mAP50-95 0.596


- Переход на более мощную модель + агрессивные аугментации повысил mAP50-95 (0.596 vs 0.582), но слегка снизил mAP50.  
- Высокий Precision и умеренный Recall → модель консервативна в детекции, упуская часть объектов.  

### Трансформер, но с приколом

In [20]:
import sys, subprocess
def ensure(pkgs): subprocess.check_call([sys.executable, "-m", "pip", "install", "--quiet", *pkgs])
ensure(["ultralytics==8.3.130"])

Используется RT-DETR-x (Transformer-based detector) — ~67 M параметров, ~232 GFLOPs при разрешении 512×512.  

**Данные:**  
Тот же Roboflow “helmet-detection” (2 класса: «With Helmet» / «Without Helmet»).  

**Тренинг:**  
– 24 эпохи  
– batch-size 16, imgsz = 512  
– начальный lr₀=2e-4 (AdamW → auto≈1.67e-3)  
– AMP, сильные аугментации (`strong_aug`) (цвет, геометрия, mosaic, mixup)  
– patience=100  

In [23]:
from ultralytics import RTDETR
YOLO = RTDETR 

rtdetrX, metrics_rtdetrX = train_and_eval(
    'rtdetr-x.pt',
    'rtdetrX_strong',
    epochs=EPOCHS_IMPROVED,
    batch=16,
    lr=2e-4,
    aug=strong_aug
)

New https://pypi.org/project/ultralytics/8.3.130 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/home/Helmet-Detection_YOLOv8-3/data.yaml, degrees=15, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=24, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=512, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.0002, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.2, mode=train, model=rtdetr-x.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=rtdetrX_strong5, nbs=64, nm

train: Scanning /home/Helmet-Detection_YOLOv8-3/train/labels.cache... 3648 images, 13 backgrounds, 0 corrupt: 100%|██████████| 3648/3648 [00:00<?, ?it/s]


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 653.2±226.9 MB/s, size: 77.5 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/valid/labels.cache... 372 images, 1 backgrounds, 0 corrupt: 100%|██████████| 372/372 [00:00<?, ?it/s]


Plotting labels to ABOBUS/rtdetrX_strong5/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.0002' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 193 weight(decay=0.0), 256 weight(decay=0.0005), 276 bias(decay=0.0)
Image sizes 512 train, 512 val
Using 6 dataloader workers
Logging results to ABOBUS/rtdetrX_strong5
Starting training for 24 epochs...

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       1/24      17.3G     0.8135      1.574     0.2945         64        512: 100%|██████████| 228/228 [01:30<00:00,  2.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.86it/s]


                   all        372        660      0.765      0.741       0.74      0.392

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       2/24      17.3G     0.6022     0.6179     0.1821         57        512: 100%|██████████| 228/228 [01:27<00:00,  2.61it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.94it/s]


                   all        372        660      0.761      0.734      0.773      0.443

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       3/24      17.3G     0.6014     0.6156     0.1857         69        512: 100%|██████████| 228/228 [01:26<00:00,  2.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  5.26it/s]


                   all        372        660      0.543      0.244      0.233      0.117

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       4/24      17.3G     0.6012      0.617     0.1854         68        512: 100%|██████████| 228/228 [01:24<00:00,  2.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.84it/s]


                   all        372        660      0.841       0.81      0.845      0.486

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       5/24      17.3G      0.583     0.6089     0.1737         83        512: 100%|██████████| 228/228 [01:20<00:00,  2.83it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.78it/s]


                   all        372        660      0.817      0.799      0.821      0.461

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       6/24      17.3G     0.5539     0.5878     0.1619         39        512: 100%|██████████| 228/228 [01:21<00:00,  2.80it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.87it/s]


                   all        372        660      0.835      0.856      0.864      0.459

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       7/24      17.3G     0.5216     0.5747      0.156         67        512: 100%|██████████| 228/228 [01:23<00:00,  2.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.70it/s]


                   all        372        660      0.844      0.863       0.88      0.502

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       8/24      17.3G     0.5273     0.5662     0.1532         39        512: 100%|██████████| 228/228 [01:24<00:00,  2.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.86it/s]

                   all        372        660      0.845      0.835      0.869      0.492



      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
       9/24      17.3G      0.514     0.5527     0.1537         85        512: 100%|██████████| 228/228 [01:24<00:00,  2.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.81it/s]


                   all        372        660      0.832      0.879      0.891      0.521

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      10/24      17.3G     0.5023      0.543     0.1435         43        512: 100%|██████████| 228/228 [01:23<00:00,  2.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.83it/s]


                   all        372        660       0.87      0.885      0.912       0.52

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      11/24      17.3G     0.4927     0.5351     0.1416         48        512: 100%|██████████| 228/228 [01:26<00:00,  2.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.79it/s]

                   all        372        660      0.863       0.87       0.89      0.516



      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      12/24      17.3G     0.4923     0.5382     0.1409         54        512: 100%|██████████| 228/228 [01:22<00:00,  2.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.80it/s]


                   all        372        660       0.89      0.853      0.903      0.502

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      13/24      17.3G     0.4833     0.5378     0.1419         53        512: 100%|██████████| 228/228 [01:25<00:00,  2.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.70it/s]


                   all        372        660      0.886      0.851      0.885      0.549

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      14/24      17.3G       0.48     0.5241     0.1327         61        512: 100%|██████████| 228/228 [01:25<00:00,  2.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.64it/s]


                   all        372        660      0.888      0.883      0.911      0.533
Closing dataloader mosaic

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      15/24      17.3G     0.3969     0.4748      0.143         33        512: 100%|██████████| 228/228 [01:25<00:00,  2.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.60it/s]


                   all        372        660       0.85      0.849      0.867      0.454

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      16/24      17.3G     0.3934     0.4613     0.1413         27        512: 100%|██████████| 228/228 [01:23<00:00,  2.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.78it/s]


                   all        372        660      0.871      0.881       0.91      0.506

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      17/24      17.3G     0.3855      0.447     0.1388         26        512: 100%|██████████| 228/228 [01:26<00:00,  2.64it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.71it/s]


                   all        372        660      0.876      0.875      0.915      0.519

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      18/24      17.3G     0.3715     0.4368     0.1319         26        512: 100%|██████████| 228/228 [01:27<00:00,  2.61it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.68it/s]


                   all        372        660      0.867      0.877      0.905      0.542

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      19/24      17.3G     0.3538     0.4273     0.1244         30        512: 100%|██████████| 228/228 [01:26<00:00,  2.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.75it/s]


                   all        372        660      0.847      0.882      0.904      0.555

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      20/24      17.3G     0.3541     0.4171     0.1231         29        512: 100%|██████████| 228/228 [01:25<00:00,  2.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.70it/s]


                   all        372        660      0.881        0.9      0.923      0.542

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      21/24      17.3G     0.3444     0.4079     0.1241         25        512: 100%|██████████| 228/228 [01:25<00:00,  2.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.63it/s]


                   all        372        660      0.875      0.856      0.903      0.565

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      22/24      17.3G     0.3369     0.4025     0.1172         37        512: 100%|██████████| 228/228 [01:26<00:00,  2.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.70it/s]


                   all        372        660       0.86      0.902      0.917      0.559

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      23/24      17.3G     0.3194     0.3934     0.1099         26        512: 100%|██████████| 228/228 [01:25<00:00,  2.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.67it/s]


                   all        372        660      0.868        0.9      0.924       0.55

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size


  0%|          | 0/228 [00:00<?, ?it/s]grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:83.)
      24/24      17.3G     0.3158     0.3847     0.1057         26        512: 100%|██████████| 228/228 [01:25<00:00,  2.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:02<00:00,  4.72it/s]


                   all        372        660      0.872      0.902      0.918      0.554

24 epochs completed in 0.616 hours.
Optimizer stripped from ABOBUS/rtdetrX_strong5/weights/last.pt, 135.4MB
Optimizer stripped from ABOBUS/rtdetrX_strong5/weights/best.pt, 135.4MB

Validating ABOBUS/rtdetrX_strong5/weights/best.pt...
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
rt-detr-x summary: 370 layers, 65,471,546 parameters, 0 gradients, 222.5 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 12/12 [00:03<00:00,  3.10it/s]


                   all        372        660      0.875      0.856      0.903      0.565
           With Helmet        215        338      0.903      0.857       0.92      0.616
        Without Helmet        208        322      0.847      0.854      0.887      0.514
Speed: 0.1ms preprocess, 4.0ms inference, 0.0ms loss, 0.4ms postprocess per image
Results saved to ABOBUS/rtdetrX_strong5
Ultralytics 8.3.129 🚀 Python-3.10.0 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100-PCIE-40GB, 40442MiB)
rt-detr-x summary: 370 layers, 65,471,546 parameters, 0 gradients, 222.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 804.4±289.5 MB/s, size: 69.7 KB)


val: Scanning /home/Helmet-Detection_YOLOv8-3/test/labels.cache... 149 images, 0 backgrounds, 0 corrupt: 100%|██████████| 149/149 [00:00<?, ?it/s]

val: /home/Helmet-Detection_YOLOv8-3/test/images/BikesHelmets446_png.rf.1dc30fe7d9690524fe33ffc1ee665fac.jpg: 1 duplicate labels removed



                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:03<00:00,  3.00it/s]


                   all        149        273        0.9      0.843      0.898      0.562
           With Helmet        110        170      0.908      0.841      0.908      0.602
        Without Helmet         61        103      0.892      0.845      0.888      0.522
Speed: 1.5ms preprocess, 13.4ms inference, 0.0ms loss, 0.4ms postprocess per image
Results saved to ABOBUS/rtdetrX_strong52
rtdetrX_strong ➜ mAP50 0.898 | mAP50-95 0.562


RT-DETR-x с сильными аугментациями показывает чуть более высокую точность (mAP@0.5) по сравнению с YOLOv8-m+strong (0.906→0.898), при этом COCO-mAP немного ниже (0.596→0.562). Transformer-модель обладает более консервативным, но стабильным детектированием — отличная альтернатива CNN в задачах с детекцией шлемов.  

## Результаты

In [53]:
scores = {
    'YOLOv8n'   : metrics_y8n,
    'YOLOv8x'   : metrics_r18,
    'YOLOv8m+'  : metrics_y8m,
    'RT-DETR-X' : metrics_rtdetrX,
#    'TinySSD'   : (miou_tiny, 0),
#    'TinySSD+'  : (miou_plus, 0),
}

print('\n### mAP50')
for name, (m50, _) in scores.items():
    print(f'{name:12s}: {m50:.3f}')

base = scores['YOLOv8n'][0]
best = max(scores, key=lambda k: scores[k][0])
gain = 100 * (scores[best][0] - base) / base

print(f'\nЛучшая модель — {best}  (+{gain:.1f}% mAP50)')


### mAP50
YOLOv8n     : 0.913
YOLOv8x     : 0.888
YOLOv8m+    : 0.906
RT-DETR-X   : 0.898

Лучшая модель — YOLOv8n  (+0.0% mAP50)


| Модель      | mAP@0.5 |
| ----------- | ------- |
| **YOLOv8n**   | **0.913** |
| YOLOv8x     | 0.888   |
| YOLOv8m+    | 0.906   |
| RT-DETR-X   | 0.898   |

- **YOLOv8n** (самая компактная) продемонстрировала лучший результат — 0.913.  
- Увеличение масштаба до YOLOv8x далось ценой падения точности (0.888).  
- Улучшенная версия YOLOv8m+ с сильными аугментациями показала 0.906, но уступила YOLOv8n.  
- RT-DETR-X набрал 0.898 — близко к YOLOv8m+, но ниже топа.  

**Вывод:**  
Лёгкая и быстрая YOLOv8n оказалась оптимальным выбором: наилучшее качество при минимальном «весе» модели.